### ISSUE TITLE
Using with clause in coforall loop for distributed k-means

### ISSUE THREAD

**Issue by novoselrok** [chapel-lang/chapel#12006]
### Summary of Problem

I was implementing a distributed memory version of the k-means algorithm. The source code is here: https://gist.github.com/novoselrok/44c2999aa9d3700f4f0642881dfa1ecc

When I try to compile the code (with the `--devel` flag) I get the following error:
`internal error: assertion error [callDestructors.cpp:681]`

I managed to reproduce it with a smaller example with the same error:
```
var arr: [{0..#numLocales}] int;

coforall L in Locales with (+ reduce arr) {
    on L {
        arr = here.id;
    }
}

writeln(arr);
```

The error seems to different based on the env variables. If I compile with `export CHPL_COMM=gasnet` and `export CHPL_COMM_SUBSTRATE=udp`, I get a different compiler error: `internal error: assertion error [createTaskFunctions.cpp:226]`

Here is the online compilation of the above code: https://tio.run/##LYzNCsIwEITveYoFLxVp8GzwDfoG4mFJtiawJrL9O5S8ujGRzmmYb2asxw9xKSsKoMgNHvtV61Nc3kOyyDTlJ4Q4G6VsGpMgMww1gIPCFmYP3QWE3GKpXZxhV1CVaumwTZXAHTwJ6eDMP84qK7VJmIlj15amlK8dGV9T6XtHK/EP

Using legacy new and default unmanaged seems to produce another different compiler error: https://tio.run/##LY5NCsIwEIXX5hQDbioyxbXFG3gD6WJIpj84TWSatkjp1Y2puH3vez@2oxdLSjMpkOoVHuulLI9@Gu7BkvC41dD7WBljQxOUROCeBfi7sPSxg@IMym6yvFecwAUImYDVHB60w1mtgeAGHSuXvavMZsyifWTxxR6pUvrYRqgdE6LjOT9CfGkexjwjYyT7xOCRVYNmS7gl@0bPy49vaJKIkx/IU8vuCw

**Compile command:**
`chpl --devel -o foo foo.chpl`

### Configuration Information

- Output of `chpl --version`:
chpl version 1.18.0
Copyright (c) 2004-2018, Cray Inc.  (See LICENSE file for more details)

- Output of `$CHPL_HOME/util/printchplenv --anonymize`:

```
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: gnu
CHPL_TARGET_ARCH: unknown
CHPL_LOCALE_MODEL: flat
CHPL_COMM: gasnet *
  CHPL_COMM_SUBSTRATE: udp *
  CHPL_GASNET_SEGMENT: everything
CHPL_TASKS: qthreads
CHPL_LAUNCHER: amudprun
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: intrinsics
  CHPL_NETWORK_ATOMICS: none
CHPL_GMP: gmp
CHPL_HWLOC: hwloc
CHPL_REGEXP: re2
CHPL_AUX_FILESYS: none
```

- Back-end compiler and version, e.g. `gcc --version` or `clang --version`:

```
gcc (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

```

**Comment by bradcray** [comment]
Hi @novoselrok — Thanks for filing this.  @vasslitvinov, would you take a look at this and see if it falls into your court?

**Comment by novoselrok** [comment]
Just as a side note, I did manage to find a workaround with a more manual approach:

```chapel
var newClustersPerLocale: [0..#numLocales][clustersDomain] Cluster;
...
newClustersPerLocale[here.id][minIndex].addPoint(point);
...
var newClusters = + reduce newClustersPerLocale;
```

**Comment by vasslitvinov** [comment]
@novoselrok - as another workaround, could you try replacing curly braces with "do" in your `coforall` loop ? I.e.
```chpl
coforall L in Locales with (+ reduce newClusters) do   // was: { .... }
```

I am working to remove the need for this workaround.

**Comment by vasslitvinov** [comment]
@novoselrok - thank you for testing the workaround. A bug fix is coming up on master...

I like the idea of using a replicated array. However I am suspicious that this update:
```chpl
newClustersPerLocale[minIndex].addPoint(point);
```
may have data races on `newClustersPerLocale[minIndex]`.

As an alternative to locking or atomics, you could use `+ reduce` in the forall loop. Probably the same as you did for the `coforall` version:

```chpl
var newClusters : ...; // perhaps [clustersSpace] Cluster

forall (point, label_) in zip(points, labels) with (+ reduce newClusters) {
  ...
  newClusters[minIndex].addPoint(point); // no sync needed: guaranteed to access task-private newClusters
}
```

**Comment by vasslitvinov** [comment]
@novoselrok I fixed these bugs on master (hopefully) in #12159.

Please let us know if you are OK to close this issue.

**Comment by novoselrok** [comment]
Yes, it seems to be ok.
